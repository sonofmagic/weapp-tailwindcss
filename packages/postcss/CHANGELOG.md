# @weapp-tailwindcss/postcss

## 1.0.0-alpha.1

### Major Changes

- [`bafd149`](https://github.com/sonofmagic/weapp-tailwindcss/commit/bafd149f0510b30cbf95711223583055023e7875) Thanks [@sonofmagic](https://github.com/sonofmagic)! - chore: monorepo changes

### Patch Changes

- Updated dependencies [[`bafd149`](https://github.com/sonofmagic/weapp-tailwindcss/commit/bafd149f0510b30cbf95711223583055023e7875)]:
  - @weapp-tailwindcss/shared@1.0.0-alpha.1

## 0.0.1-alpha.0

### Patch Changes

- Updated dependencies [[`d50d95d`](https://github.com/sonofmagic/weapp-tailwindcss/commit/d50d95d04e1c6b7c3ff32acc0d9894d3c0f06d22)]:
  - @weapp-tailwindcss/shared@0.0.1-alpha.0
