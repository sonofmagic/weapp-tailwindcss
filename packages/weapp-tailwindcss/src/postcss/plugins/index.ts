import type { AcceptedPlugin } from 'postcss'
// @ts-expect-error
import postcssIsPseudoClass from '@csstools/postcss-is-pseudo-class'
import postcssRem2rpx from 'postcss-rem-to-responsive-pixel'
import type { IStyleHandlerOptions } from '../../types'
import { postcssWeappTailwindcssPrePlugin } from './pre'
import { postcssWeappTailwindcssPostPlugin } from './post'
import { createContext } from './ctx'

export function getPlugins(options: IStyleHandlerOptions) {
  const ctx = createContext()
  options.ctx = ctx
  const plugins: AcceptedPlugin[] = [
    ...(options.postcssOptions?.plugins ?? []),
    postcssWeappTailwindcssPrePlugin(options),
    postcssIsPseudoClass({
      specificityMatchingName: 'weapp-tw-ig',
    }),
  ]
  if (options.rem2rpx) {
    plugins.push(
      postcssRem2rpx(
        typeof options.rem2rpx === 'object'
          ? options.rem2rpx
          : {
              rootValue: 32,
              propList: ['*'],
              transformUnit: 'rpx',
            },
      ),
    )
  }
  plugins.push(postcssWeappTailwindcssPostPlugin(options))
  return plugins
}

export { postcssWeappTailwindcssPostPlugin } from './post'
export { postcssWeappTailwindcssPrePlugin } from './pre'
// @ts-expect-error
export { default as postcssIsPseudoClass } from '@csstools/postcss-is-pseudo-class'
export { default as postcssRem2rpx } from 'postcss-rem-to-responsive-pixel'
