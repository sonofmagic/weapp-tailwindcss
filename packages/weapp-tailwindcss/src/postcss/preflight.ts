import type { CssPreflightOptions, IPropValue } from '../types'

export type InjectPreflight = () => IPropValue[]

export function createInjectPreflight(options?: CssPreflightOptions): InjectPreflight {
  const result: IPropValue[] = []
  // if options false ,do no thing
  if (options && typeof options === 'object') {
    const entries = Object.entries(options)
    for (const [prop, value] of entries) {
      if (value !== false) {
        result.push({
          prop,
          value: value.toString(),
        })
      }
    }
  }

  return () => {
    return result
  }
}
