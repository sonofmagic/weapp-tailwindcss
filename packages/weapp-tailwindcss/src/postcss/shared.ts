import { SimpleMappingChars2String, escape } from '../escape'
import type { InternalCssSelectorReplacerOptions } from '../types'
// css 中，要多加一个 '\' 来转义
// for raw css selector
// export function cssSelectorReplacer(selector: string, escapeEntries = MappingChars2StringEntries) {
//   return escape(selector, true, escapeEntries).replace(/\\2c /g, dic[','])
// }

export function internalCssSelectorReplacer(
  selectors: string,
  options: InternalCssSelectorReplacerOptions = {
    escapeMap: SimpleMappingChars2String,
  },
) {
  const { mangleContext, escapeMap } = options
  if (mangleContext) {
    selectors = mangleContext.cssHandler(selectors)
  }
  return escape(selectors, {
    map: escapeMap,
  })
}

export function composeIsPseudo(strs: string | string[]) {
  if (typeof strs === 'string') {
    return strs
  }
  if (strs.length > 1) {
    return `:is(${strs.join(',')})`
  }
  return strs.join('')
}
