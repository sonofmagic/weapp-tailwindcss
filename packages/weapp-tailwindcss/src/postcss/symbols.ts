export const VariablesScopeSymbol = Symbol('VariablesScope')
