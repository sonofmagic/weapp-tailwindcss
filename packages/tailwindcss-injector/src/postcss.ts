import type { Input, PluginCreator } from 'postcss'
import type { Config } from 'tailwindcss'
import type { Options } from './types'
import postcss from 'postcss'
import tailwindcss from 'tailwindcss'
import { loadConfig } from 'tailwindcss-config'
import { getConfig } from './config'
import { postcssPlugin } from './constants'
import { regExpTest, removeFileExtension } from './utils'
import { getDepFiles } from './wxml'

const configCache = new Map<string, Promise<Config | undefined>>()

async function loadTailwindConfigOnce(cwd: string, configPath: string) {
  const cacheKey = `${cwd}::${configPath}`
  let cached = configCache.get(cacheKey)
  if (!cached) {
    cached = loadConfig({
      cwd,
      config: configPath,
    }).then(result => result?.config as Config | undefined)
    configCache.set(cacheKey, cached)
  }
  return cached
}

function cloneConfig(config: Config | undefined): Config | undefined {
  if (!config) {
    return config
  }

  return {
    ...config,
  }
}

async function resolveTailwindConfig(config: Options['config'], cwd: string, sourceInput: Input | undefined) {
  if (!config) {
    return undefined
  }

  const resolved = typeof config === 'function' ? config(sourceInput) : config

  if (typeof resolved === 'string') {
    const loaded = await loadTailwindConfigOnce(cwd, resolved)
    return cloneConfig(loaded)
  }

  return cloneConfig(resolved as Config)
}

// function isObject(obj: any): obj is object {
//   return typeof obj === 'object' && obj !== null
// }
export type {
  Options,
}

const creator: PluginCreator<Partial<Options>> = (options) => {
  const { config, filter, directiveParams, cwd, extensions, insertAfterAtRulesNames, insertAfterComments } = getConfig(options)

  const extensionsGlob = `{${extensions.join(',')}}`

  return {
    postcssPlugin,
    plugins: [
      {
        postcssPlugin: `${postcssPlugin}:post`,
        async Once(root, helpers) {
          const sourceInput = root.source?.input
          if (filter(sourceInput)) {
            const nodes = root.nodes ?? []
            let atruleAnchorIndex = -1
            let commentAnchorIndex = -1
            const directivePresence = new Set<string>()

            nodes.forEach((node, index) => {
              if (node.type === 'atrule') {
                const nameMatches = insertAfterAtRulesNames.includes(node.name)
                if (nameMatches) {
                  atruleAnchorIndex = Math.max(atruleAnchorIndex, index)
                }
                if (typeof node.params === 'string') {
                  directivePresence.add(node.params)
                }
              }
              else if (node.type === 'comment' && regExpTest(insertAfterComments, node.text)) {
                commentAnchorIndex = Math.max(commentAnchorIndex, index)
              }
            })

            const anchorIndex = Math.max(atruleAnchorIndex, commentAnchorIndex)
            const anchorNode = anchorIndex >= 0 ? nodes[anchorIndex] : undefined
            let lastInserted = anchorNode

            for (const params of directiveParams) {
              if (!directivePresence.has(params)) {
                const atRule = helpers.atRule({ name: 'tailwind', params })
                if (lastInserted) {
                  root.insertAfter(lastInserted, atRule)
                }
                else {
                  root.prepend(atRule)
                }
                lastInserted = atRule
                directivePresence.add(params)
              }
            }

            if (sourceInput?.file) {
              const tailwindcssConfig = (await resolveTailwindConfig(config, cwd, sourceInput)) ?? { content: [] }

              const basename = removeFileExtension(sourceInput.file)
              const contentEntries = [`${basename}.${extensionsGlob}`]

              const deps = await getDepFiles(`${basename}.wxml`)
              for (const dep of deps) {
                contentEntries.push(`${removeFileExtension(dep)}.${extensionsGlob}`)
              }

              const uniqueEntries = Array.from(new Set(contentEntries))
              const { content: existingContent } = tailwindcssConfig as { content?: Config['content'] }

              if (!existingContent || Array.isArray(existingContent)) {
                tailwindcssConfig.content = uniqueEntries
              }
              else if (typeof existingContent === 'object' && existingContent !== null) {
                const normalizedContent = { ...existingContent }
                const currentFiles = normalizedContent.files
                const normalizedFiles = Array.isArray(currentFiles)
                  ? currentFiles
                  : typeof currentFiles === 'string'
                    ? [currentFiles]
                    : []
                normalizedContent.files = Array.from(new Set([...normalizedFiles, ...uniqueEntries]))
                tailwindcssConfig.content = normalizedContent
              }
              else {
                tailwindcssConfig.content = uniqueEntries
              }

              await postcss([
                tailwindcss(tailwindcssConfig),
              ]).process(root, {
                from: sourceInput.file,
              }).async()
            }
          }
        },
      },
    ],
  }
}
creator.postcss = true

export default creator
