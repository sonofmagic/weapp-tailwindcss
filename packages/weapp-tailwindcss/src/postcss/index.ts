import postcss from 'postcss'
import type { IStyleHandlerOptions } from '../types'
import { defuOverrideArray } from '../utils'
import { getPlugins } from './plugins'

export function styleHandler(rawSource: string, options: IStyleHandlerOptions) {
  return postcss(getPlugins(options))
    .process(
      rawSource,
      options.postcssOptions?.options ?? {
        from: undefined,
      },
    )
    .async()
}

export function createStyleHandler(options: Partial<IStyleHandlerOptions>) {
  return (rawSource: string, opt?: Partial<IStyleHandlerOptions>) => {
    return styleHandler(
      rawSource,
      defuOverrideArray<IStyleHandlerOptions, Partial<IStyleHandlerOptions>[]>(opt as IStyleHandlerOptions, options),
    )
  }
}
