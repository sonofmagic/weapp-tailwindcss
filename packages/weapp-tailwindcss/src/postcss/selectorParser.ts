import selectorParser from 'postcss-selector-parser'
import type { SyncProcessor } from 'postcss-selector-parser'
import type { Rule } from 'postcss'
import { composeIsPseudo, internalCssSelectorReplacer } from './shared'
import type { IStyleHandlerOptions } from '@/types'

const createRuleTransform = (rule: Rule, options: IStyleHandlerOptions) => {
  const { escapeMap, mangleContext, cssSelectorReplacement, cssRemoveHoverPseudoClass } = options

  const transform: SyncProcessor = (selectors) => {
    selectors.walk((selector) => {
      // do something with the selector
      // node.selector.replace(/\*/g, 'view')
      if (selector.type === 'universal' && cssSelectorReplacement && cssSelectorReplacement.universal) {
        selector.value = composeIsPseudo(cssSelectorReplacement.universal)
      }

      if (cssRemoveHoverPseudoClass && selector.type === 'selector') {
        const node = selector.nodes.find((x) => x.type === 'pseudo' && x.value === ':hover')
        node && selector.remove()
      }

      if (
        selector.type === 'pseudo' &&
        selector.value === ':root' &&
        cssSelectorReplacement &&
        cssSelectorReplacement.root
      ) {
        selector.value = composeIsPseudo(cssSelectorReplacement.root)
      }

      if (selector.type === 'class') {
        selector.value = internalCssSelectorReplacer(selector.value, {
          escapeMap,
          mangleContext
        })
      }
    })
    if (selectors.length === 0) {
      rule.remove()
    }
  }
  return transform
}

const getRuleTransformer = (rule: Rule, options: IStyleHandlerOptions) => {
  return selectorParser(createRuleTransform(rule, options))
}

export const ruleTransformSync = (rule: Rule, options: IStyleHandlerOptions) => {
  const transformer = getRuleTransformer(rule, options)

  return transformer.transformSync(rule, {
    lossless: false,
    updateSelector: true
  })
}

export function isOnlyBeforeAndAfterPseudoElement(node: Rule) {
  let b = false
  let a = false

  selectorParser((selectors) => {
    selectors.walkPseudos((s) => {
      if (s.parent?.length === 1) {
        if (/^:?:before$/.test(s.value)) {
          b = true
        }
        if (/^:?:after$/.test(s.value)) {
          a = true
        }
      }
    })
  }).astSync(node)

  return b && a
}

export const fallbackRemove = selectorParser((selectors) => {
  selectors.walk((selector) => {
    if (selector.type === 'universal') {
      selector.parent?.remove()
    }
    if (selector.type === 'pseudo' && selector.value === ':is') {
      selector.parent?.remove()
    }
  })
})
