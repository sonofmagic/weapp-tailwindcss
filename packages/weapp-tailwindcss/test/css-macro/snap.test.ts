import path from 'node:path'
import postcss from 'postcss'
import { cssCasePath, createGetCase } from '#test/util'
import creator from '@/css-macro/postcss'
const getCase = createGetCase(path.resolve(cssCasePath, 'css-macro'))

describe('snap', () => {
  it('uni-app vue2 app.css', async () => {
    const { css } = await postcss([creator]).process(await getCase('app.css'), {
      from: undefined
    })
    expect(css).toMatchSnapshot()
  })

  it('uni-app vue2 page.css', async () => {
    const { css } = await postcss([creator]).process(await getCase('page.css'), {
      from: undefined
    })
    expect(css).toMatchSnapshot()
  })

  it('uni-app vue2 app-with-comment.css', async () => {
    const { css } = await postcss([creator]).process(await getCase('app-with-comment.css'), {
      from: undefined
    })
    expect(css).toMatchSnapshot()
  })
})
