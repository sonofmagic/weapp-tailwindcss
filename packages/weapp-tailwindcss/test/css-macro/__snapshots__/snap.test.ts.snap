// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snap > uni-app vue2 app.css 1`] = `
"
/* #ifdef MP-WEIXIN */
  .ifdef-_MP-WEIXIN_cbg-blue-500{
          --tw-bg-opacity: 1;
          background-color: rgb(59 130 246 / var(--tw-bg-opacity))
  }
/* #endif */
  /* #ifndef MP-WEIXIN */
  .ifndef-_MP-WEIXIN_cbg-red-500{
          --tw-bg-opacity: 1;
          background-color: rgb(239 68 68 / var(--tw-bg-opacity))
  }
  /* #endif */
  /* #ifdef MP-WEIXIN */
  .wxcbg-blue-500{
          --tw-bg-opacity: 1;
          background-color: rgb(59 130 246 / var(--tw-bg-opacity))
  }
  /* #endif */
  /* #ifndef MP-WEIXIN */
  .-wxcbg-red-500{
          --tw-bg-opacity: 1;
          background-color: rgb(239 68 68 / var(--tw-bg-opacity))
  }
  /* #endif */
  "
`;

exports[`snap > uni-app vue2 app-with-comment.css 1`] = `
"/*!*********************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************!*\\
  !*** css ./node_modules/@vue/cli-service/node_modules/css-loader/dist/cjs.js??clonedRuleSet-22[0].rules[0].use[1]!./node_modules/@dcloudio/vue-cli-plugin-uni/packages/vue-loader/lib/loaders/stylePostLoader.js!./node_modules/@dcloudio/vue-cli-plugin-uni/packages/webpack-preprocess-loader/index.js??clonedRuleSet-22[0].rules[0].use[2]!./node_modules/postcss-loader/dist/cjs.js??clonedRuleSet-22[0].rules[0].use[3]!./node_modules/@dcloudio/vue-cli-plugin-uni/packages/sass-loader/dist/cjs.js??clonedRuleSet-22[0].rules[0].use[4]!./node_modules/@dcloudio/vue-cli-plugin-uni/packages/webpack-preprocess-loader/index.js??clonedRuleSet-22[0].rules[0].use[5]!./node_modules/@dcloudio/vue-cli-plugin-uni/packages/vue-loader/lib/index.js??vue-loader-options!./node_modules/@dcloudio/webpack-uni-mp-loader/lib/style.js!./src/pages/index/index.vue?vue&type=style&index=0&lang=scss& ***!
  \\*********************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************/
  @charset "UTF-8";
  /**
   * 这里是uni-app内置的常用样式变量
   *
   * uni-app 官方扩展插件及插件市场（https://ext.dcloud.net.cn）上很多三方插件均使用了这些样式变量
   * 如果你是插件开发者，建议你使用scss预处理，并在插件代码中直接使用这些变量（无需 import 这个文件），方便用户通过搭积木的方式开发整体风格一致的App
   *
   */
  /**
   * 如果你是App开发者（插件使用者），你可以通过修改这些变量来定制自己的插件主题，实现自定义主题功能
   *
   * 如果你的项目同样使用了scss预处理，你也可以直接在你的 scss 代码中使用如下变量，同时无需 import 这个文件
   */
  .test{
      height: 100px !important;
      width: 100px !important;
      align-items: center !important;
      justify-content: center !important;
      border-radius: 40px !important;
      background-color: rgb(18 52 86 / var(--tw-bg-opacity)) !important;
      --tw-bg-opacity: 0.54 !important;
      --tw-text-opacity: 1 !important;
      color: rgb(255 255 255 / var(--tw-text-opacity)) !important;
      display: flex !important
  }
  /* #ifdef MP-WEIXIN */
  .apply-class-0{
          --tw-bg-opacity: 1;
          background-color: rgb(59 130 246 / var(--tw-bg-opacity))
  }
  /* #endif */
  /* #ifndef MP-WEIXIN */
  .apply-class-0{
          --tw-bg-opacity: 1;
          background-color: rgb(239 68 68 / var(--tw-bg-opacity))
  }
  /* #endif */
  /* #ifdef MP-WEIXIN */
  .apply-class-1{
          --tw-bg-opacity: 1;
          background-color: rgb(59 130 246 / var(--tw-bg-opacity))
  }
  /* #endif */
  /* #ifndef MP-WEIXIN */
  .apply-class-1{
          --tw-bg-opacity: 1;
          background-color: rgb(239 68 68 / var(--tw-bg-opacity))
  }
  /* #endif */
  "
`;

exports[`snap > uni-app vue2 page.css 1`] = `
"
/* #ifdef MP-WEIXIN */
  .apply-class-0{
          --tw-bg-opacity: 1;
          background-color: rgb(59 130 246 / var(--tw-bg-opacity))
  }
/* #endif */
  /* #ifndef MP-WEIXIN */
  .apply-class-0{
          --tw-bg-opacity: 1;
          background-color: rgb(239 68 68 / var(--tw-bg-opacity))
  }
  /* #endif */
  /* #ifdef MP-WEIXIN */
  .apply-class-1{
          --tw-bg-opacity: 1;
          background-color: rgb(59 130 246 / var(--tw-bg-opacity))
  }
  /* #endif */
  /* #ifndef MP-WEIXIN */
  .apply-class-1{
          --tw-bg-opacity: 1;
          background-color: rgb(239 68 68 / var(--tw-bg-opacity))
  }
  /* #endif */
  "
`;
